packages/auto-id/src/certificateManager.ts

```diff
@@ -0,0 +1,350 @@
+
+    const derBuffer = csr.rawData
+    const signature = await crypto.subtle.sign(privateKey.algorithm.name, privateKey, derBuffer)
+    csr.signature = new Uint8Array(signature)
+
+    return csr
+  }
+
+  async createAndSignCSR(subjectName: string): Promise<x509.Pkcs10CertificateRequest> {
+    const csr = await this.createCSR(subjectName)
+    return this.signCSR(csr)
+  }
+
+  // TODO: later on move to "keyManagement.ts"
+  private static publicKeyToKeyObject(publicKey: x509.PublicKey): KeyObject {
+    // Export the key data to ArrayBuffer
+    const keyData = publicKey.rawData // DER format
+
+    // Create a KeyObject from the key data
+    const keyObject = createPublicKey({
+      key: Buffer.from(keyData),
+      format: 'der',
+      type: 'spki',
+    })
+
+    return keyObject
+  }
+
+  async issueCertificate(
+    csr: x509.Pkcs10CertificateRequest,
+    validityPeriodDays: number = 365,
+  ): Promise<x509.X509Certificate> {
+    const privateKey = this.privateKey
+    const publicKey = this.publicKey
+    if (!privateKey || !publicKey) {
+      throw new Error('Private or public key is not set.')
+    }
+
+    let issuerName: x509.Name
+    let autoId: string
+    const certificate = this.certificate
+    if (!certificate) {
+      issuerName = csr.subjectName
+      const subjectCommonName = CertificateManager.getSubjectCommonName(issuerName)
+      if (!subjectCommonName) {
+        throw new Error('Subject common name not found in CSR.')
+      }
+      autoId = blake2b_256(stringToUint8Array(subjectCommonName))
+    } else {
+      if (
+        // FIXME: modify
+        !doPublicKeysMatch(
+          CertificateManager.publicKeyToKeyObject(certificate.publicKey),
+          pemToPublicKey(await cryptoKeyToPem(publicKey)),
+        )
+      ) {
+        throw new Error(
+          'Issuer certificate public key does not match the private key used for signing.',
+        )
+      }
+
+      issuerName = certificate.subjectName
+      const certificateAutoId = CertificateManager.getCertificateAutoId(certificate) || ''
+      const certificateSubjectCommonName =
+        CertificateManager.getSubjectCommonName(certificate.subjectName) || ''
+      if (certificateAutoId === '' || certificateSubjectCommonName === '') {
+        throw new Error(
+          'Issuer certificate does not have either an auto ID or a subject common name or both.',
+        )
+      }
+      autoId = blake2b_256(
+        concatenateUint8Arrays(
+          stringToUint8Array(certificateAutoId),
+          stringToUint8Array(certificateSubjectCommonName),
+        ),
+      )
+    }
+
+    // Prepare the certificate builder with information from the CSR
+    const notBefore = new Date()
+    const notAfter = new Date()
+    notAfter.setDate(notBefore.getDate() + validityPeriodDays)
+
+    let certificateBuilder = await x509.X509CertificateGenerator.create({
+      serialNumber: randomSerialNumber().toString(),
```

Comment:
You shouldn't need to generate a random serial number here. If omitted, the [x509 package](https://peculiarventures.github.io/x509/interfaces/X509CertificateCreateWithKeyParams.html#serialNumber) will do it for you.

Reply:
I see.
Actually, i was using a custom function for this, while trying out the other libs for CertificateManager.
Removed now.